docs/src/adr/023adr-trace-evaluation.md

```diff
@@ -0,0 +1,75 @@
+# ADR-023: Trace evaluation
+
+| authors                                | last revised    |
+| -------------------------------------- | --------------: |
+| Jure Kukovec                           | 2022-09-28      |
+
+**Table of Contents**
+
+- [Summary (Overview)](#summary)
+- [Context (Problem)](#context)
+- [Options (Alternatives)](#options)
+- [Solution (Decision)](#solution)
+- [Consequences (Retrospection)](#consequences)
+
+## Summary
+
+<!-- Statement to summarize, following the following formula: -->
+
+In the context of improving usability\
+facing difficulties understanding counterexample traces\
+we decided to implement trace evaluation\
+to achieve a better user experience\
+accepting the development costs.\
+
+## Context
+
+<!-- Communicates the forces at play (technical, political, social, project).
+     This is the story explaining the problem we are looking to resolve.
+-->
+As explained in [#1845](https://github.com/informalsystems/apalache/issues/1845), one often runs into the problem of unreadable counterexamples; 
+for complex specifications, it is often the case that either the sate contains many variables, or the invariant contains many conjunctive components (or both).
+In that case, trying to determine exactly why e.g. `A1 /\ ... /\ An` was violated basically boils down to manually evaluating each formula `Ai` with whatever variables are present in the state.
+This is laborious and error-prone, and should be automated.
+
+
+## Options
+
+<!-- Communicate the options considered.
+     This records evidence of our circumspection and documents the various alternatives
+     considered but not adopted.
+-->
+1. Call REPL in each state:
+    - No convenient REPL integration at the moment
+    - No clear way of saving outputs
+1. Encode trace traversal as an Apalache problem and use the solver
+    - No additional rules or IO needed
+
+## Solution
+
+<!-- Communicates what solution was decided, and it is expected to solve the
+     problem. -->
+
+We choose option (2). 
+
+Suppose we are given a trace `s0 -> s1 -> ... -> sn` over variables `x1,...,xk` as well as expressions `E1,...,Em`, such that all free variables of `E1,...,Em` are among `x1,...,xk`. W.l.o.g. assume all constants are inlined.
+
+The above defines a trace `t0 -> t1 -> ... -> tn` over variables `v1,...,vm`, such that 
+```
+ti.vj = Ej[si.x1/x1,...,si.xk/xk]
+```
+
+for all `i \in 0..n, j \in 1..m`. In other words, `vj` in state `ti` is the evaluation of the expression `Ej` in state `si`.
+
+By using TransitionExecutor instead of a generic Next-decomposition, this can be encoded as a specification free of control-nondeterminism, in-state computation, or invariants, and is thus incredibly efficient to represent in SMT.
```

Comment:
The transition executor is very much in the guts of the model checker. While it is possible to use it, I do not think it is the simplest way of solving the problem.

Why cannot we do the following. Assume we have the following trace:

```tla
State0 ==
  /\ x = 2
  /\ y = 3

State1 ==
  /\ x = 4
  /\ y = 5

State2 ==
  /\ x = 6
  /\ y = 4
```

We could encode the above trace as a sequence of records:

```tla
Trace == <<
  [ x |-> 2, y |-> 3],
  [ x |-> 4, y |-> 5],
  [ x |-> 6, y |-> 4]
>>
```

Next, suppose we want to compute the value of `x + y` in every state of the trace. This could be done by producing the following spec:

```tla
\* the trace as defined above
Trace == << ... >>

VARIABLE
  \* the result of evaluating x + y
  \* @type: Int;
  eval1,
  \* The state number in the trace
  \* @type: Int;
  index

Init ==
  /\ eval1 = Trace[1].x + Trace[1].y
  /\ stateNo = 1

Next ==
  LET nextIndex == index + 1 IN
  /\ index' = nextIndex
  /\ eval1 = Trace[nextIndex].x + Trace[nextIndex].y
```

Note that the above specification can be produced by a preprocessing step. There is no need for direct interaction with the transition executor.

Reply:
You could do that, but then you're actually solving additional sequence and record applications in SMT. The fact that the executor is "in the guts of the model checker" is irrelevant; the user doesn't need to know or care. 
If your claim is that the executor is too difficult for Apalache developers to use, however, that is an entirely separate discussion.

Reply:
> If your claim is that the executor is too difficult for Apalache developers to use, however, that is an entirely separate discussion.

Why? This is an ADR. We can discuss the implications of development costs and long-term support :)

Reply:
> Why? This is an ADR. We can discuss the implications of development costs and long-term support :)

I think you have to write your ADRs either assuming the rest of the code is what you've got, OR, write an ADR on how to refactor/remove TEX _first_. Otherwise you get scope creep and your ADRs inevitably become a series of cascading changes and refactoring.

Reply:
I've explored it a bit, and I suppose we can keep this high level by using transition filtering instead:
```
0->0|1->1|...|n->n
```
should give us the i-th Next in the i-th step while allowing us to just call the model checker pass w/o disassembling it.

Reply:
If you're using trace exploration, user-supplied transition filtering makes no sense; there is exactly one transition between each two given states on a trace.

Reply:
> user-supplied transition filtering makes no sense

Any other tuning option then. Currently, tuning options are just a serialized String->String map.

Reply:
Trace evaluation isn't really typical BMC (it just uses it as a backend). The user should not be providing any additional parameters at all, besides the trace file and a list of expressions.

Reply:
Maybe you can explain what "_By using TransitionExecutor instead of a generic Next-decomposition, this can be encoded as a specification_" actually means.
It would be good to include such a specification, as an example. It's hard to argue about things that aren't explained in the ADR.

Reply:
I actually have a prototype, and I'll demo it tomorrow. Afterwards, we can update the ADR.

Reply:
Also, this design has been discussed offline for at least 2-3 months before the ADR.

Reply:
I think a lot of the discussion happening here has nothing to do with _design_, but is nitpicking implementation details.

Reply:
> the thing being discussed in a very specific implementation detail, and the prototype has uncontroversial stuff

Then demoing the prototype won't change the discussion here.

Reply:
I also think it's worth keeping in mind that our use of ADRs are kind of overloaded. iiuc, they were principally intended to be architectural **decision** records: and in that function it makes sense to me we may want a light-weight prototype to help inform our decisions. But we have also often used them as design documents, and perhaps this dual use can cause some confusion w/r/t expectations?

Reply:
> > Also, this design has been discussed offline for at least 2-3 months before the ADR.
> 
> Apparently not to a point where there's consensus on how to steer the transition exploration. Documenting all options on the table and weighing their benefits is exactly the point of an ADR.

But there _is_ consensus on how to steer transition exploration. Take `Next0` in the initial state, then `Next1`, etc.

The different options are about facilitating this choice of transitions (either via TRex calls or via tuning parameters, which get internally propagated into TRex calls).

Reply:
> The different options are about facilitating this choice of transitions (either via TRex calls or via tuning parameters, which get internally propagated into TRex calls).

Or via preprocessing without using TRex ;-)

Reply:
> Or via preprocessing without using TRex ;-)

Name one advantage this has over the solution with `tuning`.

Reply:
Plus, `Trace[nextIndex]` dodges all preprocessing and, most importantly, const simplification.

Reply:
> they were principally intended to be architectural **decision** records

Sure, but where do we discuss a proposed decision, if not on the PR merging the decision record?

> in that function it makes sense to me we may want a light-weight prototype to help inform our decisions

Sure, as long as such a prototype informs decisions and does not preempt them :)

Reply:
> Of course, as long as such a prototype informs decisions and does not preempt them :)

Pretty much the only tradeoff/downside of my proposed approach is _potential_ dev/maintenance cost. I find it pointless to argue how complicated an abstract implementation would be to maintain/extend. If you have a prototype in front of you, you can start making claims about which parts are maintainable and which aren't.